### PR TITLE
ddl:parallel ddl job execution problem

### PR DESCRIPTION
**Pull request by ciscoxll**
## What have you changed? (mandatory)
This pr fix problem, ddl execution in parallel will be different than expected.
two sql executed in parallel, sql1 first and sq2 later.
```
sql1: `alter table t rename to t_add;`
sql2: `alter table t add a1 int;`

expected results `[schema:1146]Table 'test_db_state.t' doesn't exist`
obtained results `IsNil`
```

sql1 and sql2 were executed successfully,the result was not expected.

Add tests

* `TestParallelRenameAddColumn`
* `TestParallelRenameDropColumn`
* `TestParallelRenameModifyColumn`
* `TestParallelRenameDropTable`
*  `TestParallelRenameCreateIndex`
* `TestParallelRenameAlterColumn`
* `TestParallelRenameRebaseAutoID`
* `TestParallelRenameTableRenameIndex`
* `TestParallelRenameTableModifyTableComment`

## What are the type of the changes (mandatory)?

Bug fix.

## How has this PR been tested (mandatory)?

Unit tests.

### CODE REVIEW

**Comment by shenli** [comment]
Please add more detail to the description.

File: model/ddl.go [hunk 1]
```diff
@@ -107,12 +107,13 @@ func (h *HistoryInfo) Clean() {
...
-	Type     ActionType    `json:"type"`
-	SchemaID int64         `json:"schema_id"`
-	TableID  int64         `json:"table_id"`
-	State    JobState      `json:"state"`
-	Error    *terror.Error `json:"err"`
+	ID        int64         `json:"id"`
+	Type      ActionType    `json:"type"`
+	SchemaID  int64         `json:"schema_id"`
+	TableID   int64         `json:"table_id"`
+	TableName string        `json:"table_name"`
```
Comment (shenli) [inline model/ddl.go]: Can we get the tablename by tableid?

File: ddl/table.go [hunk 2]
```diff
@@ -380,6 +388,34 @@ func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName
 	return nil
 }
 
+func checkOriginalTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName string) error {
```
Comment (shenli) [inline ddl/table.go]: Add comment for the function.

File: model/ddl.go (hunk 1 above)
Comment (ciscoxll) [reply model/ddl.go]: @shenli  `rename table` `tableid` has not been changed, so record the table name.

File: ddl/table.go [hunk 3]
```diff
@@ -380,6 +388,23 @@ func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName
 	return nil
 }
 
+// checkOriginalTableNotExists check if the table name is the original table name.
+// Mainly used in tableid has not changed, tablename has changed.
```
Comment (shenli) [inline ddl/table.go]: The comment is not clear.

File: ddl/table.go [hunk 4]
```diff
@@ -380,6 +388,21 @@ func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName
 	return nil
 }
 
+// checkTableNameChange checks whether the table name has been changed.
```
Comment (shenli) [inline ddl/table.go]: The comment is not clear. It provides nothing more than the function name.

Comment (shenli) [reply ddl/table.go]: Please describe why we need this checking and when will the `change` happen.

**Comment by ciscoxll** [comment]
@shenli @zimulala  @winkyao PTAL

**Comment by shenli** [comment]
@ciscoxll Please fix the CI.

**Comment by shenli** [comment]
@ciscoxll Please fix the CI.

**Comment by ciscoxll** [comment]
@zimulala @winkyao PTAL

File: ddl/table.go [hunk 5]
```diff
@@ -26,6 +26,7 @@ import (
 	"github.com/pingcap/tidb/table"
 	"github.com/pingcap/tidb/tablecodec"
 	log "github.com/sirupsen/logrus"
+	"strings"
```
Comment (zimulala) [inline ddl/table.go]: Please put it to line 18.

File: ddl/index.go [hunk 6]
```diff
@@ -349,7 +352,6 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	if err != nil {
 		return ver, errors.Trace(err)
 	}
-
```
Comment (zimulala) [inline ddl/index.go]: Do we need to add `checkTableNameChange ` here?

File: ddl/table.go (hunk 5 above)
Comment (ciscoxll) [reply ddl/table.go]: Done.

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@zimulala @winkyao PTAL

File: model/ddl.go (hunk 1 above)
Comment (zimulala) [reply model/ddl.go]: Can we put this information to `Args`?

**Comment by crazycs520** [comment]
LGTM

**Comment by zimulala** [comment]
Can we put the information of `TableName` to Args?

**Comment by shenli** [comment]
@ciscoxll Please address the comment.

File: ddl/table.go [hunk 7]
```diff
@@ -380,6 +389,24 @@ func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName
...
+// checkTableNameChange checks whether the table name has been changed, in cases where both sql1 and sql2 execute in parallel,
+// if sql1 is a renameTable it may execute successfully first, and sql2 may get the old table name, then this is not the desired result,
+// executing the renameTable DDL job does not change the tableID ,sql1 and sql2 were executed successfully,the result was not expected,
+// for example: sql1: "alter table t rename to t_add;"  sql2: "alter table t add a1 int;".
+func checkTableNameChange(t *meta.Meta, job *model.Job, tableName string) error {
+	dbInfo, err := t.GetDatabase(job.SchemaID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(job.TableName) != 0 && len(tableName) != 0 {
```
Comment (winkyao) [inline ddl/table.go]: Add some comments here, about compatibility.

File: ddl/table.go [hunk 8]
```diff
@@ -380,6 +396,26 @@ func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName
...
+// if sql1 is a renameTable it may execute successfully first, and sql2 may get the old table name, then this is not the desired result,
+// executing the renameTable DDL job does not change the tableID ,sql1 and sql2 were executed successfully,the result was not expected,
+// for example: sql1: "alter table t rename to t_add;"  sql2: "alter table t add a1 int;".
+func checkTableNameChange(t *meta.Meta, job *model.Job, tableName string, originalTableName string) error {
+	dbInfo, err := t.GetDatabase(job.SchemaID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// the old version of tidb will not check this error and will execute successfully,
+	// the new version of tidb checks for errors and prompts for errors and cancels job.
```
Comment (winkyao) [inline ddl/table.go]: "The originalTableName will be nil when the job is enqueued by old version TiDB"

File: ddl/ddl_api.go [hunk 9]
```diff
@@ -1252,7 +1251,7 @@ func (d *ddl) DropColumn(ctx sessionctx.Context, ti ast.Ident, colName model.CIS
 		TableID:    t.Meta().ID,
 		Type:       model.ActionDropColumn,
 		BinlogInfo: &model.HistoryInfo{},
-		Args:       []interface{}{colName},
+		Args:       []interface{}{colName, ti},
```
Comment (winkyao) [inline ddl/ddl_api.go]: Why not just pass the tableName string?

File: ddl/ddl_api.go [hunk 10]
```diff
@@ -1582,7 +1581,7 @@ func (d *ddl) AlterColumn(ctx sessionctx.Context, ident ast.Ident, spec *ast.Alt
 		TableID:    t.Meta().ID,
 		Type:       model.ActionSetDefaultValue,
 		BinlogInfo: &model.HistoryInfo{},
-		Args:       []interface{}{col},
+		Args:       []interface{}{col, ident.Name.O},
```
Comment (winkyao) [inline ddl/ddl_api.go]: Why not pass ident?

File: ddl/ddl_api.go [hunk 11]
```diff
@@ -1669,6 +1668,7 @@ func (d *ddl) DropTable(ctx sessionctx.Context, ti ast.Ident) (err error) {
 		TableID:    tb.Meta().ID,
 		Type:       model.ActionDropTable,
 		BinlogInfo: &model.HistoryInfo{},
+		Args:       []interface{}{ti.Name.O},
```
Comment (winkyao) [inline ddl/ddl_api.go]: ditto

File: ddl/ddl_api.go [hunk 12]
```diff
@@ -1792,7 +1792,7 @@ func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, unique bool, ind
 		TableID:    t.Meta().ID,
 		Type:       model.ActionAddIndex,
 		BinlogInfo: &model.HistoryInfo{},
-		Args:       []interface{}{unique, indexName, idxColNames, indexOption},
+		Args:       []interface{}{unique, indexName, idxColNames, indexOption, ti.Name.O},
```
Comment (winkyao) [inline ddl/ddl_api.go]: ditto

File: ddl/column.go [hunk 13]
```diff
@@ -379,11 +392,14 @@ func doModifyColumn(t *meta.Meta, job *model.Job, newCol *model.ColumnInfo, oldN
 	return ver, nil
 }
 
-func updateColumn(t *meta.Meta, job *model.Job, newCol *model.ColumnInfo, oldColName *model.CIStr) (ver int64, _ error) {
+func updateColumn(t *meta.Meta, job *model.Job, newCol *model.ColumnInfo, oldColName *model.CIStr, originalTableName string) (ver int64, _ error) {
```
Comment (winkyao) [inline ddl/column.go]: Change originalTableName from string to *model.CIStr type.

File: ddl/column.go [hunk 14]
```diff
@@ -379,11 +392,14 @@ func doModifyColumn(t *meta.Meta, job *model.Job, newCol *model.ColumnInfo, oldN
...
 	tblInfo, err := getTableInfo(t, job, job.SchemaID)
 	if err != nil {
 		return ver, errors.Trace(err)
 	}
+	if err = checkTableNameChange(t, job, tblInfo.Name.O, originalTableName); err != nil {
```
Comment (winkyao) [inline ddl/column.go]: Table name in TiDB is case insensitive, so please compare it by using tblInfo.Name.L

File: ddl/table.go [hunk 15]
```diff
@@ -380,6 +397,26 @@ func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName
 	return nil
 }
 
+// checkTableNameChange checks whether the table name has been changed, in cases where both sql1 and sql2 execute in parallel,
+// if sql1 is a renameTable it may execute successfully first, and sql2 may get the old table name, then this is not the desired result,
+// executing the renameTable DDL job does not change the tableID ,sql1 and sql2 were executed successfully,the result was not expected,
+// for example: sql1: "alter table t rename to t_add;"  sql2: "alter table t add a1 int;".
+func checkTableNameChange(t *meta.Meta, job *model.Job, tableName string, ti *ast.Ident) error {
+	dbInfo, err := t.GetDatabase(job.SchemaID)
```
Comment (zimulala) [inline ddl/table.go]: Could we put this code to line413? `dbInfo` is used only when there is an error. We can remove this operation of `get`

**Comment by zimulala** [comment]
Do we need to add this check to  other functions？Such as `onRebaseAutoID`,  `onRenameIndex`.

**Comment by zimulala** [comment]
Do we need to add this check to other functions？Such as `onDropTable`, `onTruncateTable`, and so on. Please check all DDL statements. 
And If we need to check a lot of DDL statements, it may be better to add a field to the `job` structure.

**Comment by zz-jason** [comment]
@ciscoxll PTAL the comment from @zimulala

**Comment by zimulala** [comment]
@ciscoxll Any update？

**Comment by zimulala** [comment]
@ciscoxll Any update？

**Pull request closed without merging**